pkg/controller/metrics/custom_collectors.go

```diff
@@ -328,19 +334,40 @@ func (cc clusterSyncFailingCollector) Collect(ch chan<- prometheus.Metric) {
 		return
 	}
 
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err = cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+		return
+	}
+
 	for _, cs := range clusterSyncList.Items {
 		// Failing cluster syncs
 		cond := controllerutils.FindCondition(cs.Status.Conditions, hiveintv1alpha1.ClusterSyncFailed)
 		if cond != nil && cond.Status == corev1.ConditionTrue {
-			seconds := time.Since(cond.LastTransitionTime.Time).Seconds()
-			// check if duration crosses the threshold
-			if cc.minDuration.Seconds() <= seconds {
-				ch <- prometheus.MustNewConstMetric(
-					cc.metricClusterSyncFailingSeconds,
-					prometheus.GaugeValue,
-					seconds,
-					cs.Namespace+"/"+cs.Name,
-				)
+			for _, cd := range clusterDeployments.Items {
+				// Expect to find the corresponding ClusterDeployment. In the rare case it got garbage collected by the time
+				// we fetched the list of ClusterSyncs, do not report the metric.
+				if cd.Namespace == cs.Namespace {
+					// Ignore uninstalled ClusterDeployments as it wouldn't have all conditions initialized.
+					if unreachableCondition := controllerutils.FindCondition(cd.Status.Conditions, hivev1.UnreachableCondition); unreachableCondition != nil {
+						fixedLabels := make(map[string]string, len(cc.dynamicLabels.fixedLabels))
+						fixedLabels["namespaced_name"] = cs.Namespace + "/" + cs.Name
+						fixedLabels["unreachable"] = string(unreachableCondition.Status)
+						labelValues := cc.dynamicLabels.buildLabelSlice(fixedLabels, &cd)
+						seconds := time.Since(cond.LastTransitionTime.Time).Seconds()
+						// check if duration crosses the threshold
+						if cc.minDuration.Seconds() <= seconds {
+							ch <- prometheus.MustNewConstMetric(
+								cc.metricClusterSyncFailingSeconds,
+								prometheus.GaugeValue,
+								seconds,
+								labelValues...,
+							)
+						}
+					}
+				}
+				continue
```

Comment:
redundant?

Reply:
A mistake actually, it was supposed to be `break`, to stop the looping through all the remaining cluster deployments once the appropriate one has already been found.